library/Freckle/App/Http/Cache.hs

```diff
@@ -0,0 +1,287 @@
+-- response if we receive a @304@ response.
+--
+-- - <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag#caching_of_unchanged_resources>
+--
+httpCached
+  :: forall m t
+   . MonadIO m
+  => HttpCacheSettings m t
+  -> (Request -> m (Response BSL.ByteString))
+  -> Request
+  -> m (Response BSL.ByteString)
+httpCached settings doHttp req =
+  maybe (doHttp req) handleCachableRequest $ getCachableRequestKey settings req
+ where
+  handleCachableRequest key = do
+    now <- settings.getCurrentTime
+    result <- fromEx Nothing $ settings.cache.get key
+
+    case result of
+      Nothing -> writeCache now key =<< getResponse req
+      Just val ->
+        case settings.codec.deserialise req val of
+          Left err -> do
+            settings.logWarn err
+            writeCache now key =<< getResponse req
+          Right cresp | isCachedResponseStale cresp now -> do
+            case lookupHeader hETag cresp.response of
+              Nothing -> do
+                fromEx () $ settings.cache.evict key
+                writeCache now key =<< getResponse req
+              Just etag -> do
+                resp <- getResponse $ addRequestHeader hIfNoneMatch etag req
+                case statusCode (getResponseStatus resp) of
+                  304 -> gunzipResponseBody req cresp.response
+                  _ -> do
+                    fromEx () $ settings.cache.evict key
+                    writeCache now key resp
+          Right cresp -> gunzipResponseBody req cresp.response
+
+  getResponse :: Request -> m (Response (PotentiallyGzipped BSL.ByteString))
+  getResponse = requestPotentiallyGzipped doHttp
+
+  writeCache
+    :: UTCTime
+    -> CacheKey
+    -> Response (PotentiallyGzipped BSL.ByteString)
+    -> m (Response BSL.ByteString)
+  writeCache now key resp = do
+    for_ (getCachableResponseTTL settings resp) $ \ttl -> do
+      let cresp = CachedResponse {response = resp, inserted = now, ttl = ttl}
+      fromEx () $ settings.cache.set key $ settings.codec.serialise cresp
+
+    gunzipResponseBody req resp
+
+  fromEx :: a -> m (Either SomeException a) -> m a
+  fromEx a f = do
+    result <- f
+    case result of
+      Left ex -> a <$ settings.logWarn (displayException ex)
+      Right v -> pure v
+
+-- | Return a 'CacheKey' for a 'Request', if it's cacheable
+--
+-- A 'Request' is cacheable if all are true:
+--
+-- - The given predicate succeeds
+-- - The method is @GET@
+-- - A @Cache-Control@ header with @no-store@ is not present
+--
+-- If cacheable, the 'CacheKey' is built from: method, scheme, host, port, path,
+-- query + any @Vary@ headers.
+getCachableRequestKey
+  :: HttpCacheSettings m t -> Request -> Maybe CacheKey
+getCachableRequestKey settings req = do
+  guard $ settings.cacheable req
+  guard $ HTTP.method req == "GET"
+  guard $ NoStore `notElem` requestHeaders.cacheControl
+  guard $ not settings.shared || Private `notElem` requestHeaders.cacheControl
```

Comment:
I love how much standard stuff we're getting from HTTP here, I've never appreciated HTTP as much as I do now.

Reply:
It's almost like it was well-designed! At least in this area...